Add option for playblasting from the camera sequencer.

scottdmilner opened the pull request:
Added option for setting the `sequenceTime` boolean on the playblast function. When `sequenceTime` is set to true, playblasting will ignore the camera passed to it and use the cameras from the sequencer instead. `startTime` and `endTime` are computed in sequence time instead of global time.

BigRoy commented:
Interesting @scottdmilner. Didn't know this existed, but looking at the [`maya.cmds.playblast` docs for Maya 2016](https://help.autodesk.com/cloudhelp/2016/ENU/Maya-Tech-Docs/CommandsPython/playblast.html#flagsequenceTime) this has been around a long time, which of course provides the helpful 'documentation':

> Use sequence time

Lovely.

Anyway, it looks like when using `useSequenceTime` the frames passed along should be the frames from the sequencer, not from your timeline. The sequencer may be using 0-100 but your timeline is 1001-1100. The default behavior for `capture` if no start frame and end frame are provided means to use the current timeline frame range (which, again, may not reflect the sequencer time at all).

I wonder, would it make sense to when the `useSequenceTime` is enabled to instead default to the full range of the sequencer instead. What do you think?

scottdmilner commented:
Updating the default frame range to match the sequencer range makes a lot of sense to me.

Looking at the MEL script that the Camera Sequencer window calls for playblasting the sequence, it just uses the MEL equivalent of:

```python
from maya import cmds

def get_sequencer_range():
    sequencer = cmds.sequenceManager(query=True, writableSequencer=True)
    return mc.getAttr(sequencer + ".minFrame"), mc.getAttr(sequencer + ".maxFrame")
```

---

Also, is naming the `capture` argument `sequence_time` the best option? I chose that because it matched the `cmds.playblast` flag, but it does more than just change the time, it also changes the cameras. Would it be more clear to call argument `use_sequencer` or `use_camera_sequencer`?

BigRoy commented:
> Looking at the MEL script that the Camera Sequencer window calls for playblasting the sequence, it just uses the MEL equivalent of:

Much better - thanks.

> Also, is naming the `capture` argument `sequence_time` the best option? I chose that because it matched the `cmds.playblast` flag, but it does more than just change the time, it also changes the cameras. Would it be more clear to call the argument `use_sequencer` or `use_camera_sequencer`?

I agree that Autodesk/Maya team could've picked a better argument - I'm fine with either. This may be one for @mottosso to pull a trigger on.

scottdmilner commented:
Ooh wait it should also throw an error if there is no camera sequencer node in the scene

scottdmilner commented:
Ok I've added code for setting the default frame range from the sequencer. 

---

I'm not sure that there's a good way to check for the existence of a camera sequencer, because querying it with `cmds.sequenceManager` will automatically create one if it doesn't exist. I guess you could do something with `cmds.ls`?

Would it be within the scope of this tool to check that the user provided frame range is within the start/end of the sequencer? It won't throw an error, but when I tested it, the frames from outside of the sequencer range have a visible resolution gate and aren't cropped in properly.

BigRoy commented:
> I'm not sure that there's a good way to check for the existence of a camera sequencer, because querying it with cmds.sequenceManager will automatically create one if it doesn't exist. I guess you could do something with cmds.ls?

It's nice to be clean - but I think we can ignore this one for now. That node gets auto created even when opening the Camera Sequencer UI and not doing anything else. Let's assume for now that if anyone uses the `sequence_time` argument they intentionally wanted to do so and hence would likely have a sequencer anyway.

> Would it be within the scope of this tool to check that the user provided frame range is within the start/end of the sequencer? It won't throw an error, but when I tested it, the frames from outside of the sequencer range have a visible resolution gate and aren't cropped in properly.

I'd ignore that for now.

BigRoy reviewed (approved):
Looks good to me.

scottdmilner commented:
In doing more testing, I realized that the `_applied_camera_options` context manager will not work properly because it only operates on a single camera. I will take a look at that tomorrow.

mottosso commented:
Neat. To me, the most sensical name is in the description of the argument, "Use the Camera Sequencer" -> `use_camera_sequencer=True`. 

> In doing more testing, I realized that the _applied_camera_options context manager will not work properly because it only operates on a single camera. I will take a look at that tomorrow.

Maybe we need a test scene or test case to go with this, something that produces a playblast using the sequencer and multiple cameras?

> Would it be within the scope of this tool to check that the user provided frame range is within the start/end of the sequencer?

Can we clamp min/max values given by the user to the min/max values of the sequencer? With a minimum duration of 1? My understanding is that values outside of there are non-existent and non-sensical, given the lack of a camera.

BigRoy commented:
> Can we clamp min/max values given by the user to the min/max values of the sequencer? With a minimum duration of 1? My understanding is that values outside of there are non-existent and non-sensical, given the lack of a camera.

For what it's worth - it does work outside of the range it seems, it just falls back to the nearest camera in time maybe? I haven't investigated much further though. I think force clamping it maybe doing too much (even though sure, it may be sensible for most cases)

scottdmilner commented:
I've updated the argument to be `use_camera_sequencer` and added some code to `_applied_camera_options` to get the camera options working with all cameras. 

---

It looks like when you use a start frame that's too early, it just keeps the current viewport camera and doesn't update the frame number, and then when you use an end frame that's too late it basically acts as a frame hold. There's no catastrophic failure or anything—it appear that clamping would be necessary to maintain functionality.

mottosso commented:
Looks good to me, nice work!
